src/engine/sys/sdl_glimp.cpp

```diff
@@ -1384,9 +1384,46 @@ void GLimp_HandleCvars()
 {
 	if ( r_swapInterval->modified )
 	{
-		AssertCvarRange( r_swapInterval, -1, 1, true );
+		/* Set the swap interval for the GL context.
+
+		* -1 : adaptive sync
+		* 0 : immediate updates
+		* 1 : updates synchronized with the vertical refresh
+		* N : updates syncrhonized with Nth vertical refresh
+
+		For example if screen has 60 Hz refresh rate:
+
+		* -1 will update the screen 60 times per second, using adaptive sync if supported,
+		* 0 will update the screen as soon as it can,
+		* 1 will update the screen 60 times per second,
+		* 2 will update the screen 30 times per second.
+		* 3 will update the screen 20 times per second,
+		* 4 will update the screen 15 times per second.
+
+		5 is arbitrarily set as a ceiling value to prevent
+		mistakes making the game almost unresponsive.
+
+		About adaptive sync:
+
+		> Some systems allow specifying -1 for the interval, to enable adaptive vsync.
+		> Adaptive vsync works the same as vsync, but if you've already missed the
+		> vertical retrace for a given frame, it swaps buffers immediately, which
+		> might be less jarring for the user during occasional framerate drops.
+		> If an application requests adaptive vsync and the system does not support it,
+		> this function will fail and return -1. In such a case, you should probably
+		> retry the call with 1 for the interval.
+		> -- https://wiki.libsdl.org/SDL_GL_SetSwapInterval */
+
+		AssertCvarRange( r_swapInterval, -1, 5, true );
+
 		R_SyncRenderThread();
-		SDL_GL_SetSwapInterval( r_swapInterval->integer );
+
+		if ( SDL_GL_SetSwapInterval( r_swapInterval->integer ) == -1 )
```

Comment:
According to https://wiki.libsdl.org/SDL_GL_SetSwapInterval the argument must be in [-1, 1]

Reply:
That page does not document higher numbers unfortunately. I started to dig this topic and the more I find, the more I think there may be a mistake in not documenting higher numbers… and lower number.

Because experience shows that `6` **and** `-6` not only does not produce an error, but also produce the expected behaviour: drawing a frame 1/N time the frequency of the screen, the positive/negativeness being about the _kind_ of sync.

For example that page links to the “glx_swap_control_tear.txt” file, unfortunately the given link is dead but here is the new url:

- https://www.khronos.org/registry/OpenGL/extensions/EXT/GLX_EXT_swap_control_tear.txt

This page says:

> This extension extends the existing GLX_EXT_swap_control extension by
> allowing a negative <interval> parameter to glXSwapIntervalEXT. The
> negative <interval> allows late swaps to occur without synchronization
> to the video frame. This reduces the visual stutter on late frames and
> reduces the stall on subsequent frames.

What's interesting about the `-1` is that they don't say the “minus one value” but the “negative interval”, it's not a magic number to activate this or that feature, it's an interval. In `-1`, the One is not the number One option, the One is the value One, the value between Zero and Two.

To begin with, just with this sentence, the idea of `[-1, 0, 1]` interval don't match the wording.

To know more about intervals and be sure that's what our Wiki said, I looked at the Swap Interval page on Khronos wiki (linked on that SDL page):

- https://www.khronos.org/opengl/wiki/Swap_Interval

And this page says:

> A swap interval of 1 tells the GPU to wait for one v-blank before swapping the front and back buffers. A swap interval of 0 specifies that the GPU should never wait for v-blanks, thus performing buffer swaps as soon as possible when rendering for a frame is finished.

And also says:

> A swap interval greater than 0 means that the GPU may force the CPU to wait due to previously issued buffer swaps.

So it's not about being equal to 1, but being positive. Now we know the interval is go beyond than `[-1, 0, 1]` on the positive area, can it go beyond `-1`?

I returned to the `GLX_EXT_swap_control_tear.txt` file in extension registry and it said:

> "If <interval> is negative, the minimum number of video frames between
> buffer swaps is the absolute value of <interval>.

And continue:

> In this case, if
> abs(<interval>) video frames have already passed from the previous swap
> when the swap is ready to be performed, the swap will occur without
> synchronization to a video frame. When an unsynchronized swap happens,
> the missed frame will count towards the minimum number of video frames
> for the next swap."

Which is in fact explaining the alternative sync mode.

This information put together means the absolute value always code the same thing, a value (an interval) that can be greater than 1, and the sign is a flag to select an alternate sync function.

The max value is said to be implementation-dependent:

> The current swap interval and implementation-dependent max swap
> interval for a particular drawable can be obtained by calling
> glXQueryDrawable with the attribute […]

Following those findings, I updated the PR accordingly to support wider `[-N, 0, N]` range.